### ISSUE TITLE
Identify possible algorithms for MNIST

### ISSUE THREAD

**Issue by sajit** [sajit/kaggles#1]
Brain storm & identify various options for digit recognition.
1. NN
2. SVM
3. K-means clustering?
4. Random Forest
5. Logistic Regression

**Comment by icarusin** [comment]
For 1) using scikit-learn seems the best way to start off for a quick "out of the box" implementation of NN. There is the training data + test data provided for MNIST handwritten digits already so I could use that directly to train and then run it over test data. Based on the results from scikit-learn we could decide on using theano or pybrain or anything else.

**Comment by sajit** [comment]
@icarusin  ill upload the script later but I was able to get a basic Pybrain version working. NN seem to be the most popular choice so let's  stick with that for now. 
The library to be honest doesn't matter - used Pybrain coz that was the first google result.

What I think matters is how the model is tweaked. How to train it. Would be great if you can do some research.

Right now , what I have is 

 input layer , with 784 inputs
output layer with 10 outputs (0..9)
hidden layer : 200 [just picked a random number ]
Current problems : In various training iterations I see performance actually go down.. Need to understand why error is actually going up. 

TODOS
- 1. Understand how to calculate training error &  test error
- 2. Why model is actually doing worse after 10 iterations..

**Comment by icarusin** [comment]
While I continue to grapple to get a basic NN running with scikit, here's some links on how to decide on the number of hidden layers and number of nodes per hidden layer.

ftp://ftp.sas.com/pub/neural/FAQ3.html#A_hu
ftp://ftp.sas.com/pub/neural/FAQ3.html#A_hl

**Comment by sajit** [comment]
So da @icarusin  Ive uploaded the basic 101 version I had. I hope the code is self explanatory. Outstanding issues.. I had tried 2 attempts , and I think attempt2 is what Pybrain recommends. So feel free to delete the function attempt1.
- Need to split the train data into 0.75 , 0.25 train / cross validation - so that the neural network can better understand how its doing with error in each iteration of training aka epoch
- Need to understand the differences between various layers, Sigmoid, Linear, Softmax etc. I blindly copied and pasted from the sample in Pybrain docs
- Need to research on neural networks training and understand how to train NNs such that they don't overfit and stop when they reach a threshold error rate.

**Comment by sajit** [comment]
Hopefully this gives you a starting point. . Gmail me if you're stuck. You need to install pybrain before you can run this example. Sort of a pain on windows. 
My go to guide : http://pybrain.org/docs/tutorial/fnn.html

**Comment by sajit** [comment]
@icarusin I noticed that mostly was stuck trying to read data from csv into a legible format that PyBrain could use.. and after that writing it back.. And also learning to split the dataset correctly.. Hopefully since thats done, we can start tweaking the NN itself and start playing around with different 
- training methodologies
- types of layers and neuorons

**Comment by icarusin** [comment]
@sajit Here's an implementation of MLP (what you have implemented) which has the code for softmax and prediction. I'm still getting around python so maybe you will have better understanding on this.

https://gist.github.com/amueller/2061456

**Comment by icarusin** [comment]
@sajit Since I had already got mid-way with installing scikit-learn, I have implemented the SVM Classifier for digits data from kaggle. It was straightforward. The problems with NN and SVM based classifiers boil down to the parameter selection for the "black-boxes". For SVM its this "gamma" value and the process involves couple steps like grid-search and cross-validation

http://scikit-learn.org/stable/modules/grid_search.html#grid-search
http://scikit-learn.org/stable/modules/cross_validation.html#cross-validation

I think these are generic to selecting parameters for neural network as well. Need to research more. 
Btw, if you want to get windows-binaries for python modules, this below is an excellent link:

http://www.lfd.uci.edu/~gohlke/pythonlibs/

Just have to "pip install **binary-file-name**" and its done. I installed numpy, scipy this way.
